core/drivers/crypto/hisilicon/hpre_ecc.c

```diff
@@ -0,0 +1,967 @@
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	if (BITS_TO_BYTES(size_bits) != BITS_TO_BYTES(curve->key_bits)) {
+		EMSG("Invalid size_bits %lu.", size_bits);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = gen_random_k(key->d, curve->key_bits, curve->n);
+	if (ret)
+		return ret;
+
+	ret = hpre_ecc_request_init(curve, &msg, key->d, NULL);
+	if (ret) {
+		EMSG("Ecc key pair request init fail.");
+		return ret;
+	}
+
+	ret = hpre_do_ecc_task(&msg);
+	if (ret) {
+		EMSG("Fail to do ecc key pair task! ret = 0x%x.", ret);
+		goto done;
+	}
+
+	ecc_dh = &msg.param_size.ecc_dh;
+	ret = crypto_bignum_bin2bn(msg.out, ecc_dh->rx_bytes, key->x);
+	if (ret) {
+		EMSG("Fail to trans res x to bn.");
+		goto done;
+	}
+
+	ret = crypto_bignum_bin2bn(msg.out + msg.key_bytes,
+				   ecc_dh->ry_bytes, key->y);
+	if (ret)
+		EMSG("Fail to trans res y to bn.");
+done:
+	hpre_ecc_request_deinit(&msg);
+	return ret;
+}
+
+static TEE_Result hpre_ecc_do_shared_secret(struct drvcrypt_secret_data *sdata)
+{
+	struct hpre_ecc_msg msg = { };
+	struct hpre_ecc_curve *curve = NULL;
+	struct ecc_public_key *pubkey = NULL;
+	struct ecc_keypair *ecc_key = NULL;
+	struct hpre_ecc_dh *ecc_dh = NULL;
+	TEE_Result ret = TEE_SUCCESS;
+
+	if (!sdata || !sdata->key_priv || !sdata->key_pub) {
+		EMSG("Invalid ecc_do_shared_secret input parameters.");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ecc_key = (struct ecc_keypair *)sdata->key_priv;
+	pubkey = (struct ecc_public_key *)sdata->key_pub;
+	if (!ecc_key || !pubkey) {
+		EMSG("Invalid key param.");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	curve = get_curve_from_list(ecc_key->curve);
+	if (!curve) {
+		EMSG("Fail to get valid curve, id %u.", ecc_key->curve);
+		return TEE_ERROR_NOT_SUPPORTED;
+	}
+
+	if (sdata->size_sec != BITS_TO_BYTES(curve->key_bits)) {
+		EMSG("Invalid sdata size_sec %lu.", sdata->size_sec);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = hpre_ecc_request_init(curve, &msg, ecc_key->d, pubkey);
+	if (ret) {
+		EMSG("Ecc shared secret request init fail.");
+		return ret;
+	}
+
+	ret = hpre_do_ecc_task(&msg);
+	if (ret) {
+		EMSG("Fail to do ecc shared secret task! ret = 0x%x.", ret);
+		goto done;
+	}
+
+	ecc_dh = &msg.param_size.ecc_dh;
+	/*
+	 * Only take the x coordinate as the output result, according to
+	 * soft computing implementation.
+	 */
+	memcpy(sdata->secret.data, msg.out, ecc_dh->rx_bytes);
+	sdata->secret.length = ecc_dh->rx_bytes;
+	memzero_explicit(msg.out, ECC_DH_OUT_SIZE(msg.key_bytes));
+
+done:
+	hpre_ecc_request_deinit(&msg);
+	return ret;
+}
+
+static struct drvcrypt_ecc driver_ecc = {
+	.alloc_keypair = hpre_ecc_alloc_keypair,
+	.alloc_publickey = hpre_ecc_alloc_publickey,
+	.free_publickey = hpre_ecc_free_publickey,
+	.gen_keypair = hpre_ecc_gen_keypair,
+	.shared_secret = hpre_ecc_do_shared_secret,
+};
+
+TEE_Result hpre_ecc_init(void)
+{
+	TEE_Result ret = drvcrypt_register_ecc(&driver_ecc);
+
+	if (ret != TEE_SUCCESS)
+		EMSG("Hpre ecc register to crypto fail");
+
+	return ret;
```

Comment:
In case of error, OP-TEE will only print a message and continue initialization. If you want the platform to refuse to boot (and not fallback to a software ECC implementation, you should explicitly panic here.